core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncPathCache.java

```diff
@@ -11,201 +11,348 @@
+import java.util.function.BiConsumer;
 import javax.annotation.Nullable;
 import javax.annotation.concurrent.ThreadSafe;
 
 /**
- * This cache maintains the Alluxio paths which have been synced with UFS.
+ * A cache of paths that have been synchronized or have been marked as not being synchronized with
+ * the UFS. Paths are mapped to time values that indicate their last synchronization or
+ * invalidation.
+ *
+ * Synchronization times are updated only at the root path for which a sync was performed by
+ * {@link alluxio.master.file.InodeSyncStream}.
+ * Each path contains the following values concerning synchronization:
+ * - recursiveValidationTime: the last time a sync was successful on this path that included syncing
+ * all of its children recursively
+ * - validationTime: the last time a sync was successful on this path
+ * - directValidationTime: the last time a sync was successful on the path which (at least)
+ * included itself and its direct children
+ * - recursiveValidationTime: the last time a sync was successful on this path which (at least)
+ * included itself and all its children
+ * Each path additionally contains the following values considering invalidation:
+ * - invalidationTime: the last time this exact path was invalidated
+ * - directChildInvalidation: the last time a direct child of this path was invalidated
+ * - recursiveChildInvalidation: the last time a non-direct child of this path was invalidated
+ * - isFile: when checking if a file needs to be synced, any descendant type from the most
+ * recent sync will be valid
+ *
+ * Whenever an invalidation is received the path, and its parents up to the root have their
+ * appropriate invalidation times updated. Validation times are updated on the path
+ * when {@link #notifySyncedPath} is called on the root sync path after a successful sync.
+ * An invalidation is received either when a client calls
+ * {@link alluxio.master.file.DefaultFileSystemMaster#invalidateSyncPath} to notify that a path
+ * needs synchronization, or when a file is updated by an external Alluxio cluster, and
+ * cross cluster sync is enabled.
+ *
+ * Checking if a path needs to be synchronized involves checking the appropriate validation
+ * and invalidation times at each path component up to the root.
  */
 @ThreadSafe
-public final class UfsSyncPathCache {
+public class UfsSyncPathCache {
   private static final Logger LOG = LoggerFactory.getLogger(UfsSyncPathCache.class);
 
-  /** Number of paths to cache. */
-  private static final int MAX_PATHS =
-      Configuration.getInt(PropertyKey.MASTER_UFS_PATH_CACHE_CAPACITY);
-
-  /** Cache of paths which have been synced. */
-  private final Cache<String, SyncTime> mCache;
-
+  public final Cache<String, SyncState> mItems;
   private final Clock mClock;
+  // we keep the root as a separate value as we never want to evict it,
+  // and instead we just update it under a lock
+  private final SyncState mRoot = new SyncState(false);
```

Comment:
If you want an entry to never be evicted then you might want to consider using weight instead of total size.
https://github.com/google/guava/wiki/CachesExplained#eviction
Simply put, an entry with weight=0 will NEVER be evicted. That kind of entries might help if you have a special group you never want to evict, for example, the root, mount points and other kinds of big dirs.

Reply:
I did some testing on this and found that entries with weight 0 were being evicted. Using the weight based cache also seemed to have performance implications on the effectiveness of LRU and overall speed of the cache so I decided it was not a good route to go.

Reply:
Ah that I didn't know. Thanks for checking that out. The current approach looks good to me.

Not relevant to this PR, I guess in the future we can consider holding a collection of SyncState for the directories we don't want to evict, for example the root and mount points. Then we have this Cache for all the files and small directories probably. Then maybe we no longer need to worry about evicting the state on an important dir and cause it to be invalidated. Just a thought.